### ISSUE TITLE
[Messenger] Redis transport flawed / Multiple consumers for same transport impossible

### ISSUE THREAD

**Issue by toooni** [symfony/symfony#35358]
**Symfony version(s) affected**: 4.3.0+

**Description**  
As far as I've seen it isn't possible to have two consumers on the same transport. We have the possibility to define the stream, group and consumer per transport. But none of this config options allow to have two or more consumer commands `messenger:consume` running on the same transport.

**How to reproduce** 
- Use a config similar to this:
```
framework:
    messenger:
        transports:
            async_priority_high:
                dsn: 'redis://redis-server/priority_high'
```
- Run two or more consume commands in parallel for the same transport `bin/console messenger:consume async_priority_high`
- Send a message through the `async_priority_high` transport.

Now there is the chance that one of the commands exits with an exception like:
> Symfony\Component\Messenger\Exception\TransportException: Could not acknowledge redis message "1579079882286-4".

This also means that the message was consumed twice.

~~Unfortunately this happens because the message is only being acknowledged (Redis `XACK`) after the message is handled and leaves it for the duration of the handling in the stream to be consumed by the other consumer command too.~~
EDIT: This is not true. The issue does only happen once if the two consumers are started around the same time and at least one message is already in the queue.

**Possible Solution**  
~~Using redis `XACK` right after `XREADGROUP` in `Symfony\Component\Messenger\Transport\RedisExt\Connection:get()` and using the `Symfony\Component\Messenger\Transport\RedisExt\Connection:reject()` method to re-add the message to the redis stream.~~
EDIT: With the edit above, this isn't a solution.

**Additional context**  
I am sorry if I did get the usage of messenger wrong, but IMO having multiple consume commands running at the same time should be supported.

**Comment by toooni** [comment]
@amigian74 If the transport is something like SQS or AMQP, I agree - but I think Redis streams are not supposed to solve the same problems like a "real" message queue. This means that the redis transport must be forced to behave like it's supposed to by the messenger component.
What happens if you redeploy your solution and you get a different random string for the consumer? Aren't the existing, unconsumed messages in the old consumer lost?

**Comment by alexander-schranz** [comment]
@toooni Have also look at @toflar PRs: https://github.com/symfony/symfony-docs/pull/11869/files

**Comment by toooni** [comment]
@alexander-schranz Yes, I've seen that. But this should not be necessary if you don't care which host executes which message. It is a feature we can use if we want the same host to execute only the messages originated from it. But I want both hosts working the same queue.

**Comment by alexander-schranz** [comment]
@toooni when you want todo it on the same host you need to set a unique CONSUMER_NAME for every process you start. A consumer need to be unique and you shouldn't use a random number else it would not continue correctly on crash so for example if you want to have 3 parallels consumer you do something like this:

```
CONSUMER_NAME=WORKER_1 bin/console messenger:consume ...
CONSUMER_NAME=WORKER_2 bin/console messenger:consume ...
CONSUMER_NAME=WORKER_3 bin/console messenger:consume ...
```

**Comment by toooni** [comment]
@alexander-schranz Yes, I got that. But this still doesn't solve the issue I intended to solve. Why can't I run the same consumer twice? I know redis doesn't support that since streams are as the name says, streams ;) and are intended to deliver a stream of messages to every consumer connected.

I am not 100% sure but my perception of the messenger component is that you should be able to switch the transport and you'll get the same behaviour if you use the `message:consume` command.

**Comment by alexander-schranz** [comment]
@toooni in a queue system a consumer should be unique else you will get into race conditions and streams are a queue system with groups and consumers.
I'm not sure about the other transport implementation but I personally think if the process get killed it should continue with the pending message of that consumer and not that the message should get lost, maybe thats currently the case for the other implementations but that would be in my case a bug in the other transports if that message would get lost and never consumed.

**Comment by toooni** [comment]
@alexander-schranz I am sorry if I don't got the intentions of Redis Labs right. But I used SQS and RabbitMQ and can assure you that these do not work like that. So either the messenger component does intend to deliver the same experience to all the transports or it intends to do what these transports can do. That's why I've added the following to "Additional context":
> I am sorry if I did get the usage of messenger wrong, but IMO having multiple consume commands running at the same time should be supported.

**Comment by alexander-schranz** [comment]
I invested some time and currently there is no way to workaround that you have to configure per process a unique consumer name.

The only thing we could do would be adding something like `?pendings=false` flag or something similiar to avoid reading messages which where read but not acked. This will have the disadvantage that you could lose messages as redis does not redeliver messages which where read but not acked after some time, so the message would get lost forever if the consume process crash between read and ack.

**Comment by toooni** [comment]
I just created a scenario to reproduce my issue:
- I started two consumers with the same parameter `bin/console messenger:consume async_priority_low`
- I've sent a `SleepMessage` which is handled by a `SleepHandler` which sleeps for some seconds on every message handling.

This resulted in exactly what I would expect from symfony messenger (and the redis transport). Every Message was only handled once by only one consumer.

Unfortunately if I do send the `SleepMessage` before I've started my consumers and started the consumers around the same time. The first message will be handled twice and the second handling will result in a `TransportException`:
>  [Symfony\Component\Messenger\Exception\TransportException]  
  Could not acknowledge redis message "1579246274621-0".

**Comment by rapli** [comment]
If there are multiple consumers with the same consumer name within the same group one consumer the following may happen:
message 1 => consumer A1
message 1 => consumer A2
consumer A1 processes and acknowledges message 1
consumer A2 processes message 1 and tries to acknowledge it, but it is already acknowledged

This behaviour can be reproduced even with redis-cli and can only be prevented by using distinct consumer names (using sleeps or anything else is not really safe).
If I understand you correctly you do not want a worker queue but a notification queue?

In order to allow multi consumer notifications the Redis transport should allow to not facilitate consumer groups, then reading the stream e.g. via xread (which probably would still require distinct consumer names though), so every consumer has its own view on the stream. I am not sure about the design of the messenger component, but I think sending messages to multiple consumers is a valid use case.

**Comment by toooni** [comment]
@rapli When I am talking about running the same `messenger:consume` command twice, this does not definitely mean that each of these processes then can't have their own redis consumer. I am talking about letting the redis transport deal with this issue.
I've implemented a solution with the help of @alexander-schranz in which we can resolve the issue about having to know about the number or names of the redis consumers (I will create a PR).
I've used XPENDING and XCLAIM to get messages of abandoned redis consumers. This also makes https://github.com/symfony/symfony-docs/pull/11869/files partly deprecated because messages will not be lost anymore if a consumer won't come back.

My idea was to use a unique consumer on every process to fix my issue but @alexander-schranz pointed out that this might lead to issues because of too much abandoned consumers in redis. Even if my problem isn't completely solved yet, we could at least safely use a hostname for the redis consumer name.

**Comment by toooni** [comment]
I've created a PR (https://github.com/symfony/symfony/pull/35384) to allow having the hostname as the consumer name. This would also make it possible to solve this issue for me by my own config.

Maybe I will introduce another PR to allow random consumer names by configuration.

**Comment by najmi9** [comment]
I am working symfony5.3 and I have the same  problem, I added the line
 ```yaml
#/etc/supervisor/conf.d/messenger-worker.conf
environment = REDIS_CONSUMER=consumer-%(process_num)d
```
```yaml
#config/packages/messenger.yaml
framework:
    messenger:
        transports:
            async: 
              dsn: 'redis://localhost:6379/messages/symfony/%env(REDIS_CONSUMER)%' # Env var not found
...
```
now my problem is in Symfony application the variable REDIS_CONSUMER, How I can make it defined.

**Comment by PhilETaylor** [comment]
I had this exact issue when deploying mt first messenger bus to production today, with a high workload, with 50 consumers and Redis Transport. 

using `XINFO CONSUMERS messages symfony` I confirmed that all my workers had different names (as I had already written this into my config)...

First 4 as an example:

<img width="264" alt="Screenshot 2022-01-31 at 20 31 46" src="https://user-images.githubusercontent.com/400092/151868442-e1282e66-fbd6-470c-8c43-88342678d3a5.png">

And then within seconds of a production load coming online (say 5-10 messages) Im getting the error 

```
Could not acknowledge redis message "1643660792108-0"
```

Stacktrace from sentry
https://sentry.io/share/issue/8ddea9afeafd4b029ee977cc433e4ed2/

I removed `lazy=true` from the DSN and the problem eased a little, but as I typing this I got another 10 failures. 

It works perfectly with a single consumer.. just not 50 started at once with Supervisor... 

Might need to fork this to a new issue or reopen this issue?

**Comment by PhilETaylor** [comment]
I am an idiot... 

3 x Docker containers...x50 consumers should be 150 consumers running, and only 50 listed... because all with "messenger-consume_XX" consumer names ....DUH.... Sorry for the noise... I'll go add the hostname to the env var now...
